src/core/layout/qgslayoutitemlegend.cpp

```diff
@@ -845,47 +855,119 @@ QgsExpressionContext QgsLayoutItemLegend::createExpressionContext() const
 
   context.appendScope( scope );
 
-  return context;
+  if ( replace )
+  {
+    mExpContext.~QgsExpressionContext();
+    const QList<QgsExpressionContextScope *> scopes = context.takeScopes();
+    for ( QgsExpressionContextScope *scopep : scopes )
+    {
+      mExpContext.appendScope( scopep );
+    }
+    return mExpContext;
+  }
+  else
+  {
+    return context;
+  }
 }
 
+
 // -------------------------------------------------------------------------
 #include "qgslayertreemodellegendnode.h"
 #include "qgsvectorlayer.h"
+#include "qgsmaplayerlegend.h"
 
 QgsLegendModel::QgsLegendModel( QgsLayerTree *rootNode, QObject *parent )
   : QgsLayerTreeModel( rootNode, parent )
 {
+  mLayoutLegendContext = nullptr;
   setFlag( QgsLayerTreeModel::AllowLegendChangeState, false );
   setFlag( QgsLayerTreeModel::AllowNodeReorder, true );
 }
 
+Qt::ItemFlags QgsLegendModel::flags( const QModelIndex &index ) const
+{
+  // make the legend nodes selectable even if they are not by default
+  if ( index2legendNode( index ) )
+    return QgsLayerTreeModel::flags( index ) | Qt::ItemIsSelectable;
+
+  return QgsLayerTreeModel::flags( index );
+}
+
 QVariant QgsLegendModel::data( const QModelIndex &index, int role ) const
 {
   // handle custom layer node labels
-  if ( QgsLayerTreeNode *node = index2node( index ) )
-  {
-    if ( QgsLayerTree::isLayer( node ) && ( role == Qt::DisplayRole || role == Qt::EditRole ) && !node->customProperty( QStringLiteral( "legend/title-label" ) ).isNull() )
+  QString name;
+  QgsLayerTreeNode *node = index2node( index );
+  QgsLayerTreeModelLegendNode *ltmln = index2legendNode( index ); // Possibly useless
```

Comment:
Possibly useless?

Reply:
From the test results and the debugging, it would seem that the index that are iterated over don't provide an index that can be converted into a legend node even for the layers that have a categorical symbology (the jet layer in the legend layout test notably).